subscriptions/src/subscriptions.ts

```diff
@@ -0,0 +1,214 @@
+import * as im from 'immutable';
+import { from, merge, Observable, Observer, Subject } from 'rxjs';
+import {
+  startWith,
+  concatMap,
+  flatMap,
+  filter,
+  finalize,
+  ignoreElements,
+  scan,
+  tap,
+  map,
+  publishReplay,
+  refCount,
+  share,
+} from 'rxjs/operators';
+import { applyReducer, Operation } from 'fast-json-patch';
+import {
+  KeyedPatches,
+  Patch,
+  Version,
+  Versioned,
+} from '@binaris/shift-interfaces/dist/subscriptions';
+import { mapWithState, StateAndOutput, takeUntilLast } from './rxutils';
+
+interface SubscriptionState<T> extends Versioned<T> {
+  readonly patches?: ReadonlyArray<Patch>;
+}
+
+type VersionedStateGetter<T> = (key: string) => Promise<Versioned<T>>;
+
+export const REGISTER: 'register' = 'register';
+export const DEREGISTER: 'deregister' = 'deregister';
+const UPDATE: 'update' = 'update';
+
+interface RegisterPollerInput {
+  readonly action: 'register';
+  readonly key: string;
+  readonly version: Version;
+}
+interface DeregisterPollerInput {
+  readonly action: 'deregister';
+  readonly key: string;
+}
+interface UpdatePollerInput {
+  readonly action: 'update';
+  readonly keysToVersions: Array<[string, Version]>;
+}
+type PollerInput = RegisterPollerInput | DeregisterPollerInput;
+
+export type NonEmptyArray<T> = [T, ...T[]];
+
+export function isNonEmptyArray<T>(arr: ReadonlyArray<T>): arr is NonEmptyArray<T> {
+  return arr.length > 0;
+}
+
+export type KeyedNonEmptyPatch = [string, NonEmptyArray<Patch>];
+export type KeyedNonEmptyPatches = NonEmptyArray<KeyedNonEmptyPatch>;
+export type Poller = (keysToVersions: im.Map<string, Version>) => Promise<KeyedPatches>;
+
+export function pollReducer(
+  keysToVersions: im.Map<string, Version>,
+  input: PollerInput | UpdatePollerInput,
+): im.Map<string, Version> {
+  switch (input.action) {
+    case REGISTER:
+      return keysToVersions.set(input.key, input.version);
+    case DEREGISTER:
+      return keysToVersions.delete(input.key);
+    case UPDATE:
+      // Only update existing keys in case key was updated after deregister
+      return keysToVersions.withMutations((m) =>
+        input.keysToVersions.forEach(([k, v]) => m.update(k, (prev) => prev === undefined ? prev : v))
+      );
+  }
+}
+
+export function poll(poller: Poller) {
+  const subject = new Subject<UpdatePollerInput>();
+  return (obs$: Observable<PollerInput>) => {
+    const sharedObs$ = obs$.pipe(share());
+    return merge(
+      sharedObs$,
+      subject.pipe(takeUntilLast(sharedObs$)),
+    ).pipe(
+      scan(pollReducer, im.Map()),
+      filter((keysToVersions) => keysToVersions.size > 0),
+      // TODO: backoff on error
+      concatMap(poller),
+      map((patchesPerKey) =>
+        patchesPerKey.filter((keyedPatch): keyedPatch is KeyedNonEmptyPatch => isNonEmptyArray(keyedPatch[1]))),
```

Comment:
I don't understand why look at the second element of `keyedPatch` to declare `keyedPatch` is nonempty.